File: java/filament/src/main/java/io/github/erkko68/filament/BufferObject.java
package io.github.erkko68.filament;

import io.github.erkko68.filament.internal.NativeRegistry;
import java.lang.ref.Cleaner;
import java.nio.Buffer;

public class BufferObject {
    private long mNativeObject;

    private BufferObject(long nativeBufferObject) {
        mNativeObject = nativeBufferObject;
    }

    public static class Builder {
        private final long mNativeBuilder;
        private final Cleaner.Cleanable mCleanable;

        public Builder() {
            mNativeBuilder = nCreateBuilder();
            mCleanable = NativeRegistry.registerForCleanup(this, new BuilderCleanup(mNativeBuilder));
        }

        public Builder size(int byteCount) {
            nBuilderSize(mNativeBuilder, byteCount);
            return this;
        }

        public Builder bindingType(BindingType type) {
            nBuilderBindingType(mNativeBuilder, type.ordinal());
            return this;
        }

        public BufferObject build(Engine engine) {
            long nativeObject = nBuilderBuild(mNativeBuilder, engine.getNativeObject());
            if (nativeObject == 0) throw new IllegalStateException("Couldn't create BufferObject");
            return new BufferObject(nativeObject);
        }

        private static class BuilderCleanup implements Runnable {
            private final long mNativeObject;
            BuilderCleanup(long nativeObject) { mNativeObject = nativeObject; }
            @Override public void run() { nDestroyBuilder(mNativeObject); }
        }
    }

    public enum BindingType {
        VERTEX,
        UNIFORM
    }

    public void setBuffer(Engine engine, Buffer buffer) {
        setBuffer(engine, buffer, 0, buffer.remaining());
    }

    public void setBuffer(Engine engine, Buffer buffer, int byteOffset, int byteCount) {
        nSetBuffer(getNativeObject(), engine.getNativeObject(), buffer, buffer.remaining(), byteOffset, byteCount);
    }

    public long getNativeObject() {
        if (mNativeObject == 0) {
            throw new IllegalStateException("Calling method on destroyed BufferObject");
        }
        return mNativeObject;
    }

    void clearNativeObject() {
        mNativeObject = 0;
    }

    private static native long nCreateBuilder();
    private static native void nDestroyBuilder(long nativeBuilder);
    private static native void nBuilderSize(long nativeBuilder, int byteCount);
    private static native void nBuilderBindingType(long nativeBuilder, int type);
    private static native long nBuilderBuild(long nativeBuilder, long nativeEngine);
    private static native void nSetBuffer(long nativeBufferObject, long nativeEngine, Buffer buffer, int remaining, int byteOffset, int byteCount);
}

File: kotlin/filament/src/jvmMain/kotlin/io/github/erkko68/filament/BufferObject.kt
package io.github.erkko68.filament

import java.nio.ByteBuffer
import java.nio.ByteOrder

actual class BufferObject(val nativeBufferObject: io.github.erkko68.filament.jni.BufferObject) {
    actual enum class BindingType {
        VERTEX, UNIFORM, SHADER_STORAGE;
        internal fun toJni(): io.github.erkko68.filament.jni.BufferObject.Builder.BindingType = when (this) {
            VERTEX -> io.github.erkko68.filament.jni.BufferObject.Builder.BindingType.VERTEX
            // JVM JNI bindings currently expose only VERTEX.
            else -> throw UnsupportedOperationException("BindingType $this is not supported by JVM JNI BufferObject.Builder")
        }
    }

    actual class Builder actual constructor() {
        private val jni = io.github.erkko68.filament.jni.BufferObject.Builder()

        actual fun size(byteCount: Int): Builder {
            jni.size(byteCount)
            return this
        }

        actual fun bindingType(bindingType: BindingType): Builder {
            jni.bindingType(bindingType.toJni())
            return this
        }

        actual fun build(engine: Engine): BufferObject = 
            BufferObject(jni.build(engine.nativeEngine))
    }

    actual fun getByteCount(): Int = nativeBufferObject.byteCount

    actual fun setBuffer(engine: Engine, data: ByteArray) {
        val buffer = ByteBuffer.allocateDirect(data.size).order(ByteOrder.nativeOrder())
        buffer.put(data)
        buffer.rewind()
        nativeBufferObject.setBuffer(engine.nativeEngine, buffer)
    }

    actual fun setBuffer(engine: Engine, data: ByteArray, destOffsetInBytes: Int, count: Int) {
        val buffer = ByteBuffer.allocateDirect(data.size).order(ByteOrder.nativeOrder())
        buffer.put(data)
        buffer.rewind()
        nativeBufferObject.setBuffer(engine.nativeEngine, buffer, destOffsetInBytes, count)
    }

    actual fun setBuffer(engine: Engine, data: ByteArray, destOffsetInBytes: Int, count: Int, callback: (() -> Unit)?) {
        val buffer = ByteBuffer.allocateDirect(data.size).order(ByteOrder.nativeOrder())
        buffer.put(data)
        buffer.rewind()
        val runnable: java.lang.Runnable? = if (callback != null) Runnable { callback() } else null
        nativeBufferObject.setBuffer(engine.nativeEngine, buffer, destOffsetInBytes, count, null, runnable)
    }
}
